text/0000-read-all.md

```diff
@@ -0,0 +1,99 @@
+- Feature Name: read_exact and read_full
+- Start Date: 2015-03-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Rust's `Write` trait has `write_all`, which is a convenience method that calls
+`write` repeatedly to write an entire buffer. This proposal adds two similar
+convenience methods to the `Read` trait: `read_full` and `read_exact`.
+`read_full` calls `read` repeatedly until the buffer has been filled, EOF has
+been reached, or an error other than `Interrupted` occurs. `read_exact` is
+similar to `read_full`, except that reaching EOF before filling the buffer is
+considered an error.
+
+# Motivation
+
+The `read` method may return fewer bytes than requested, and may fail with an
+`Interrupted` error if a signal is received during the call. This requires
+programs wishing to fill a buffer to call `read` repeatedly in a loop. This is
+a very common need, and it would be nice if this functionality were provided in
+the standard library. Many C and Rust programs have the same need, and solve it
+in the same way. For example, Git has [`read_in_full`][git], which behaves like
+the proposed `read_full`, and the Rust byteorder crate has
+[`read_full`][byteorder], which behaves like the proposed `read_exact`.
+[git]: https://github.com/git/git/blob/16da57c7c6c1fe92b32645202dd19657a89dd67d/wrapper.c#L246
+[byteorder]: https://github.com/BurntSushi/byteorder/blob/2358ace61332e59f596c9006e1344c97295fdf72/src/new.rs#L184
+
+# Detailed design
+
+The following methods will be added to the `Read` trait:
+
+``` rust
+fn read_full(&mut self, buf: &mut [u8]) -> Result<usize>;
```

Comment:
@ArtemGr 

> Right now any generic code must check for `EINTR` and short reads for the chance that it will be used with an I/O stream, right? But what if I'm reading from a byte slice? Should I pay the price of `EINTR` and short read checks?

That's actually part of the advantage of having `read_full` be a trait method with a default implementation. Generic code can always call `read_full` with guaranteed consistent behavior (no `Interrupted`, a short read only on EOF), and it will by default accomplish that by calling `read` in a loop. However, any `Read` implementation is free to override `read_full` if they can implement it more efficiently. For your example of reading from a byte slice, `read_full` will always behave identically to `read` so the `Read` impl for `[u8]` could override `read_full` to be a straight call to `read`. Additionally, since this version of the RFC proposes the default `read_exact` be implemented with a call to `read_full`, that would also use the simplified version for free.